### PR TITLE
MULE-16396: OnErrorPropagate should only rollback owned transactions

### PR DESCRIPTION
**Pull request by mbuchwald**

### CODE REVIEW

File: core/src/main/java/org/mule/runtime/core/internal/processor/DelegateTransaction.java [hunk 1]
```diff
@@ -216,5 +218,15 @@ public void resume() throws TransactionException {}
...
       return null;
     }
+
+    @Override
+    public ComponentLocation getComponentLocation() {
+      return null;
+    }
+
+    @Override
+    public void setComponentLocation(ComponentLocation componentLocation) {
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/processor/DelegateTransaction.java]: Code Smell Code Smell: Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. ([squid:S1186](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS1186))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7640&issues=AWilgAaSotIXnmsUpHDS&open=AWilgAaSotIXnmsUpHDS)

[ISSUE_KEY]: AWilgAaSotIXnmsUpHDS

File: core/src/main/java/org/mule/runtime/core/internal/exception/OnErrorPropagateHandler.java [hunk 2]
```diff
@@ -7,15 +7,21 @@
 package org.mule.runtime.core.internal.exception;
 
 import static reactor.core.publisher.Mono.just;
+
+import org.mule.runtime.api.component.location.ComponentLocation;
 import org.mule.runtime.core.api.event.CoreEvent;
 import org.mule.runtime.core.api.processor.Processor;
+import org.mule.runtime.core.api.transaction.Transaction;
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/exception/OnErrorPropagateHandler.java]: Code Smell Code Smell: Remove this unused import 'org.mule.runtime.core.api.transaction.Transaction'. ([squid:UselessImportCheck](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7640&issues=AWirRZ1EotIXnmsUDiqP&open=AWirRZ1EotIXnmsUDiqP)

[ISSUE_KEY]: AWirRZ1EotIXnmsUDiqP

File: core/src/main/java/org/mule/runtime/core/internal/exception/OnErrorPropagateHandler.java [hunk 3]
```diff
@@ -7,15 +7,21 @@
 package org.mule.runtime.core.internal.exception;
 
 import static reactor.core.publisher.Mono.just;
+
+import org.mule.runtime.api.component.location.ComponentLocation;
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/exception/OnErrorPropagateHandler.java]: Code Smell Code Smell: Remove this unused import 'org.mule.runtime.api.component.location.ComponentLocation'. ([squid:UselessImportCheck](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7640&issues=AWirRZ1EotIXnmsUDiqR&open=AWirRZ1EotIXnmsUDiqR)

[ISSUE_KEY]: AWirRZ1EotIXnmsUDiqR

File: core/src/main/java/org/mule/runtime/core/internal/exception/OnErrorPropagateHandler.java [hunk 4]
```diff
@@ -34,13 +40,22 @@ public boolean acceptsAll() {
...
       return event;
     };
   }
 
+  private boolean isOwnedTransaction() {
+    TransactionAdapter transaction = (TransactionAdapter) TransactionCoordination.getInstance().getTransaction();
+    if (transaction == null || !transaction.getComponentLocation().isPresent()) {
+      return false;
+    }
+    return transaction.getComponentLocation().get().getRootContainerName()
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/exception/OnErrorPropagateHandler.java]: Bug Bug: Call "Optional#isPresent()" before accessing the value. ([squid:S3655](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS3655))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7640&issues=AWirRZ1DotIXnmsUDiqO&open=AWirRZ1DotIXnmsUDiqO)

[ISSUE_KEY]: AWirRZ1DotIXnmsUDiqO

File: core/src/main/java/org/mule/runtime/core/internal/processor/TryScope.java [hunk 5]
```diff
@@ -33,15 +34,19 @@
...
 import org.mule.runtime.core.privileged.processor.Scope;
 import org.mule.runtime.core.privileged.processor.chain.MessageProcessorChain;
 
+import org.mule.runtime.core.privileged.transaction.TransactionAdapter;
 import org.reactivestreams.Publisher;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.util.List;
+import java.util.Optional;
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/processor/TryScope.java]: Code Smell Code Smell: Remove this unused import 'java.util.Optional'. ([squid:UselessImportCheck](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7640&issues=AWirRZ1rotIXnmsUDiqS&open=AWirRZ1rotIXnmsUDiqS)

[ISSUE_KEY]: AWirRZ1rotIXnmsUDiqS

File: core/src/main/java/org/mule/runtime/core/internal/exception/OnErrorPropagateHandler.java [hunk 6]
```diff
@@ -7,15 +7,21 @@
...
 import org.mule.runtime.core.api.event.CoreEvent;
 import org.mule.runtime.core.api.processor.Processor;
+import org.mule.runtime.core.api.transaction.Transaction;
+import org.mule.runtime.core.api.transaction.TransactionCoordination;
 import org.mule.runtime.core.privileged.exception.MessageRedeliveredException;
 import org.mule.runtime.core.privileged.exception.TemplateOnErrorHandler;
 
 import java.util.ArrayList;
 import java.util.List;
+import java.util.Optional;
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/exception/OnErrorPropagateHandler.java]: Code Smell Code Smell: Remove this unused import 'java.util.Optional'. ([squid:UselessImportCheck](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7640&issues=AWirRZ1EotIXnmsUDiqQ&open=AWirRZ1EotIXnmsUDiqQ)

[ISSUE_KEY]: AWirRZ1EotIXnmsUDiqQ

File: core/src/main/java/org/mule/runtime/core/internal/processor/DelegateTransaction.java [hunk 7]
```diff
@@ -216,5 +223,15 @@ public void resume() throws TransactionException {}
...
       return null;
     }
+
+    @Override
+    public Optional<ComponentLocation> getComponentLocation() {
+      return empty();
+    }
+
+    @Override
+    public void setComponentLocation(ComponentLocation componentLocation) {
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/processor/DelegateTransaction.java]: Code Smell Code Smell: Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. ([squid:S1186](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS1186))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7640&issues=AWilgAaSotIXnmsUpHDS&open=AWilgAaSotIXnmsUpHDS)

[ISSUE_KEY]: AWilgAaSotIXnmsUpHDS

File: core/src/main/java/org/mule/runtime/core/internal/exception/OnErrorPropagateHandler.java (hunk 4 above)
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/exception/OnErrorPropagateHandler.java]: Bug Bug: Call "Optional#isPresent()" before accessing the value. ([squid:S3655](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS3655))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7640&issues=AWirRZ1DotIXnmsUDiqO&open=AWirRZ1DotIXnmsUDiqO)

[ISSUE_KEY]: AWirRZ1DotIXnmsUDiqO

File: core/src/main/java/org/mule/runtime/core/internal/processor/DelegateTransaction.java (hunk 7 above)
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/processor/DelegateTransaction.java]: Code Smell Code Smell: Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. ([squid:S1186](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS1186))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7640&issues=AWilgAaSotIXnmsUpHDS&open=AWilgAaSotIXnmsUpHDS)

[ISSUE_KEY]: AWilgAaSotIXnmsUpHDS

File: core/src/main/java/org/mule/runtime/core/internal/exception/OnErrorPropagateHandler.java (hunk 2 above)
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/exception/OnErrorPropagateHandler.java]: Code Smell Code Smell: Remove this unused import 'org.mule.runtime.core.api.transaction.Transaction'. ([squid:UselessImportCheck](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7640&issues=AWirRZ1EotIXnmsUDiqP&open=AWirRZ1EotIXnmsUDiqP)

[ISSUE_KEY]: AWirRZ1EotIXnmsUDiqP

File: core/src/main/java/org/mule/runtime/core/internal/exception/OnErrorPropagateHandler.java (hunk 3 above)
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/exception/OnErrorPropagateHandler.java]: Code Smell Code Smell: Remove this unused import 'org.mule.runtime.api.component.location.ComponentLocation'. ([squid:UselessImportCheck](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7640&issues=AWirRZ1EotIXnmsUDiqR&open=AWirRZ1EotIXnmsUDiqR)

[ISSUE_KEY]: AWirRZ1EotIXnmsUDiqR

File: core/src/main/java/org/mule/runtime/core/internal/exception/OnErrorPropagateHandler.java (hunk 6 above)
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/exception/OnErrorPropagateHandler.java]: Code Smell Code Smell: Remove this unused import 'java.util.Optional'. ([squid:UselessImportCheck](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7640&issues=AWirRZ1EotIXnmsUDiqQ&open=AWirRZ1EotIXnmsUDiqQ)

[ISSUE_KEY]: AWirRZ1EotIXnmsUDiqQ

File: core/src/main/java/org/mule/runtime/core/internal/processor/TryScope.java (hunk 5 above)
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/processor/TryScope.java]: Code Smell Code Smell: Remove this unused import 'java.util.Optional'. ([squid:UselessImportCheck](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7640&issues=AWirRZ1rotIXnmsUDiqS&open=AWirRZ1rotIXnmsUDiqS)

[ISSUE_KEY]: AWirRZ1rotIXnmsUDiqS

File: core/src/main/java/org/mule/runtime/core/internal/exception/OnErrorPropagateHandler.java [hunk 8]
```diff
@@ -34,13 +37,22 @@ public boolean acceptsAll() {
...
       return event;
     };
   }
 
+  private boolean isOwnedTransaction() {
+    TransactionAdapter transaction = (TransactionAdapter) TransactionCoordination.getInstance().getTransaction();
+    if (transaction == null || !transaction.getComponentLocation().isPresent()) {
+      return false;
+    }
+    return transaction.getComponentLocation().get().getRootContainerName()
```
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/exception/OnErrorPropagateHandler.java]: Bug Bug: Call "Optional#isPresent()" before accessing the value. ([squid:S3655](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS3655))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7640&issues=AWirRZ1DotIXnmsUDiqO&open=AWirRZ1DotIXnmsUDiqO)

[ISSUE_KEY]: AWirRZ1DotIXnmsUDiqO

File: core/src/main/java/org/mule/runtime/core/internal/processor/DelegateTransaction.java (hunk 7 above)
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/processor/DelegateTransaction.java]: Code Smell Code Smell: Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. ([squid:S1186](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS1186))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7640&issues=AWilgAaSotIXnmsUpHDS&open=AWilgAaSotIXnmsUpHDS)

[ISSUE_KEY]: AWilgAaSotIXnmsUpHDS

File: core/src/main/java/org/mule/runtime/core/internal/exception/OnErrorPropagateHandler.java (hunk 8 above)
Comment (marianogonzalez) [reply core/src/main/java/org/mule/runtime/core/internal/exception/OnErrorPropagateHandler.java]: sonar is right

File: core/src/main/java/org/mule/runtime/core/privileged/transaction/AbstractTransaction.java [hunk 9]
```diff
@@ -195,4 +197,14 @@ public int getTimeout() {
...
   }
+
+  @Override
+  public void setComponentLocation(ComponentLocation componentLocation) {
+    this.componentLocation = componentLocation;
+  }
+
+  @Override
+  public Optional<ComponentLocation> getComponentLocation() {
+    return Optional.ofNullable(componentLocation);
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/privileged/transaction/AbstractTransaction.java]: static import

File: core/src/main/java/org/mule/runtime/core/privileged/transaction/TransactionAdapter.java [hunk 10]
```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.core.privileged.transaction;
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/privileged/transaction/TransactionAdapter.java]: why privileged and not internal?

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/source/SourceCallbackAdapter.java [hunk 11]
```diff
@@ -52,4 +53,5 @@
    */
   String getOwningExtensionName();
 
+  ComponentLocation getSourceLocation();
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/source/SourceCallbackAdapter.java]: jdoc

File: core/src/main/java/org/mule/runtime/core/internal/exception/OnErrorPropagateHandler.java (hunk 8 above)
Comment (mbuchwald) [reply core/src/main/java/org/mule/runtime/core/internal/exception/OnErrorPropagateHandler.java]: But the `isPresent` is invoked in the if that is shown...........

Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/exception/OnErrorPropagateHandler.java]: Bug Bug: Call "Optional#isPresent()" before accessing the value. ([squid:S3655](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS3655))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7640&issues=AWirRZ1DotIXnmsUDiqO&open=AWirRZ1DotIXnmsUDiqO)

[ISSUE_KEY]: AWirRZ1DotIXnmsUDiqO

File: core/src/main/java/org/mule/runtime/core/internal/processor/DelegateTransaction.java (hunk 7 above)
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/processor/DelegateTransaction.java]: Code Smell Code Smell: Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. ([squid:S1186](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS1186))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7640&issues=AWilgAaSotIXnmsUpHDS&open=AWilgAaSotIXnmsUpHDS)

[ISSUE_KEY]: AWilgAaSotIXnmsUpHDS

File: core/src/main/java/org/mule/runtime/core/internal/exception/OnErrorPropagateHandler.java (hunk 8 above)
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/exception/OnErrorPropagateHandler.java]: Bug Bug: Call "Optional#isPresent()" before accessing the value. ([squid:S3655](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS3655))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7640&issues=AWirRZ1DotIXnmsUDiqO&open=AWirRZ1DotIXnmsUDiqO)

[ISSUE_KEY]: AWirRZ1DotIXnmsUDiqO

File: core/src/main/java/org/mule/runtime/core/internal/processor/DelegateTransaction.java (hunk 7 above)
Comment (sonar-mulesoft) [inline core/src/main/java/org/mule/runtime/core/internal/processor/DelegateTransaction.java]: Code Smell Code Smell: Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. ([squid:S1186](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS1186))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7640&issues=AWilgAaSotIXnmsUpHDS&open=AWilgAaSotIXnmsUpHDS)

[ISSUE_KEY]: AWilgAaSotIXnmsUpHDS

**Pull request merged**